### PR TITLE
[AMD] Triton Backend for ROCm

### PR DESCRIPTION
**Pull request by micmelesse**
Hi, this is a pr to add a Triton backend to Flash Attention on ROCm. We hope that this pr will be the first in a series of prs to that end. Triton has had support for ROCm for a while now and a Flash Attention Triton backend will allow us to support Flash Attention on both our CDNA (MI200 & MI300) and RDNA Machines.

Below is the state of features in this pr.

These features are supported in Fwd and Bwd
1) Fwd and Bwd with causal masking
2) Variable sequence lengths
3) Arbitrary Q and KV sequence lengths
4) Arbitrary head sizes

These features are supported in Fwd for now. We will add them to backward soon.
1) Multi and grouped query attention
2) ALiBi

These features are in development
1) Paged Attention 
2) Sliding Window
3) Rotary embeddings
4) Dropout
5) Performance Improvements

We have created a test file, `tests/test_flash_attn_triton_amd.py` which is a subset of `tests/test_flash_attn.py`. It currently contains the following tests. The tests are the same as the main test files with some configs disabled that are not yet supported. All sequence lengths and head sizes are the same as the original.  They all pass on an MI200 machine.

1. `test_flash_attn_qkvpacked`
2. `test_flash_attn_varlen_qkvpacked`
3. `test_flash_attn_output`
4. `test_flash_attn_varlen_output`
5. `test_flash_attn_causal`
6. `test_flash_attn_varlen_causal`
7. `test_flash_attn_kvcache`

image

There is clearly more work to be done but we hope that this will make a good start. We have included instructions to run the Triton Backend in the README but the main point is to use `export FLASH_ATTENTION_TRITON_AMD_ENABLE="TRUE"` with Triton installed.

Please let us know what we can do on our end to help with this process.

Finally this pr includes work from multiple people besides myself, special thanks to @vgokhale, @scxiao and @jlgreathouse.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -61,7 +61,7 @@
 SKIP_CUDA_BUILD = os.getenv("FLASH_ATTENTION_SKIP_CUDA_BUILD", "FALSE") == "TRUE"
 # For CI, we want the option to build with C++11 ABI since the nvcr images use C++11 ABI
 FORCE_CXX11_ABI = os.getenv("FLASH_ATTENTION_FORCE_CXX11_ABI", "FALSE") == "TRUE"
-
+USE_TRITON_ROCM = os.getenv("FLASH_ATTENTION_USE_TRITON_ROCM", "FALSE") == "TRUE"
```
Comment (dtrifiro) [inline setup.py]: If `USE_TRITON_ROCM` is true, we can also skip the `git submodule update` command  at line 137, as well as as the `os.system` calls  to generate code from the composable kernel repo, line 309.

Comment (micmelesse) [reply setup.py]: done

**Comment by unclemusclez** [comment]
The Gods are Gracious

**Comment by unclemusclez** [comment]
will this work with CDNA 1?

**Comment by micmelesse** [comment]
> will this work with CDNA 1?

The kernels work on any architecture supported by the Triton compiler. Right now the Triton compiler does not officially support MI100 series but most cases should work. We are focused on MI300 and MI200 on the CDNA side.

**Comment by micmelesse** [comment]
Hi @tridao 

Hope you are doing well. I wanted to check if you have any feedback or suggestions regarding this PR. I've refreshed it to include support for the backward pass and have refactored it to be more modular and easier to review.

We would be happy to add more features or work on performance improvements if needed. If you have any fundamental reservations about adding a Triton backend, please let us know, and we will do everything we can to address them.

Thank you for your time.

**Comment by dtrifiro** [comment]
Is there anything holding this back?

**Comment by micmelesse** [comment]
> Is there anything holding this back?

We are just waiting for feedback

**Comment by TerminatorJ** [comment]
These features are supported in Fwd for now. We will add them to backward soon.
1.Multi and grouped query attention
2.ALiBi

does this mean the AMD version can not support the alibi_bias before the backwards are added? When using alibi_bias, I always get value instability as nan during the training time.

**Comment by micmelesse** [comment]
We should be putting up a pr in a week or two with backwards support and alibi. I will ping you when it is up.

**Comment by micmelesse** [comment]
Second pr is up at https://github.com/Dao-AILab/flash-attention/pull/1610 and the discussion is at https://github.com/Dao-AILab/flash-attention/pull/1561

**Pull request merged**